### ISSUE TITLE
FramePadding Injected Incorrectly While Using Groups

### ISSUE THREAD

**Issue by atom0s** [ocornut/imgui#4190]
**Version/Branch of Dear ImGui:**

Version: a640d8a6d2bfd2e8f125c1dc72594543808ecddf
Branch: master

**Back-end/Renderer/Compiler/OS**

Back-ends: imgui_impl_d3d8.cpp (custom) + imgui_impl_win32.cpp
Compiler: Visual Studio 2019 16.10.0
Operating System: Windows 10 Pro

**My Issue/Question:**

When using groups and children, it is possible to get additional groups to misalign when calling `SameLine()` between groups.

Here is an example of groups working fine while having text before/after a child within groups:

```cpp
        ImGui::SetNextWindowSize(ImVec2(400, 400), ImGuiCond_Always);
        ImGui::Begin("Test Window Bug");
        {
            // Left side..
            ImGui::BeginGroup();
            {
                ImGui::TextColored(ImVec4(1.0, 1.0, 1.0, 1.0), "Example Left Top");
                ImGui::BeginChild("left_side", ImVec2(150.0f, -ImGui::GetFrameHeightWithSpacing()));
                {}
                ImGui::EndChild();
                ImGui::TextColored(ImVec4(1.0, 1.0, 1.0, 1.0), "Example Left Bottom");
            }
            ImGui::EndGroup();

            ImGui::SameLine();

            // Right side..
            ImGui::BeginGroup();
            {
                ImGui::TextColored(ImVec4(1.0, 1.0, 1.0, 1.0), "Example Right Top");
                ImGui::BeginChild("right_side", ImVec2(150.0f, -ImGui::GetFrameHeightWithSpacing()));
                {}
                ImGui::EndChild();
                ImGui::TextColored(ImVec4(1.0, 1.0, 1.0, 1.0), "Example Right Bottom");
            }
            ImGui::EndGroup();
        }
        ImGui::End();
```

While using some controls in the 'Bottom' parts after `EndChild` is called, such as Text and TextColored, the second group is properly aligned to the same line as the first. 

![FNrPWBm](https://user-images.githubusercontent.com/1422090/120417159-3c655180-c313-11eb-8b46-17052cc4b002.png)

However, if other certain controls are used, such as a button, the second group becomes misaligned to the size of the current FramePadding.y value. For example:

```cpp
        ImGui::SetNextWindowSize(ImVec2(400, 400), ImGuiCond_Always);
        ImGui::Begin("Test Window Bug");
        {
            // Left side..
            ImGui::BeginGroup();
            {
                ImGui::TextColored(ImVec4(1.0, 1.0, 1.0, 1.0), "Example Left Top");
                ImGui::BeginChild("left_side", ImVec2(150.0f, -ImGui::GetFrameHeightWithSpacing()));
                {}
                ImGui::EndChild();
                ImGui::Button("Example Left Bottom");
            }
            ImGui::EndGroup();

            ImGui::SameLine();

            // Right side..
            ImGui::BeginGroup();
            {
                ImGui::TextColored(ImVec4(1.0, 1.0, 1.0, 1.0), "Example Right Top");
                ImGui::BeginChild("right_side", ImVec2(150.0f, -ImGui::GetFrameHeightWithSpacing()));
                {}
                ImGui::EndChild();
                ImGui::TextColored(ImVec4(1.0, 1.0, 1.0, 1.0), "Example Right Bottom");
            }
            ImGui::EndGroup();
        }
        ImGui::End();
```

![VqS8lGy](https://user-images.githubusercontent.com/1422090/120417294-75052b00-c313-11eb-9370-2be242dc9618.png)

Here you can see the top text for the right group is shifted down 3 pixels in this case, which is what my FramePadding.y value is. 

To demonstrate the issue exists using stock ImGui code and theme, here is also another picture of it running in the D3D9 example on the default theme. Just to show it's not specific to my environment/renderer. 

![IbY06Ac](https://user-images.githubusercontent.com/1422090/120417828-59e6eb00-c314-11eb-963c-9b8ff4b294d4.png)

**Comment by atom0s** [comment]
To add to this, it seems any control that is based on a button is causing this behavior to happen. 
But, if you draw a new line of text after the bottom button on the left, it'll correct the issue. (Although it'll be an undesired look of the window in general.)

This narrows down the issue into: `ImGui::ItemSize` and more specifically:

https://github.com/ocornut/imgui/blob/master/imgui.cpp#L7381
```cpp
window->DC.PrevLineTextBaseOffset = ImMax(window->DC.CurrLineTextBaseOffset, text_baseline_y);
```

When SameLine is called, this value is carried over. 

This also leads to inconsistent alignment based on which order you render certain elements.

For example, rendering a `button > sameline > text > sameline > button` all on the same line results in the text being aligned to the button's text:
![https://i.imgur.com/quNkRLK.png](https://i.imgur.com/quNkRLK.png)

Where as doing a `text > sameline > button > sameline > button` has the text above the buttons text alignment:
![https://i.imgur.com/mQC4VFF.png](https://i.imgur.com/mQC4VFF.png)

Basically anytime a button type is rendered (button, combo, checkbox, etc.) it'll result in the text following it to align differently than someone may expect. Then anytime text is drawn, the next element will be reset from the button's alignment before it which may align things unexpectedly as well.
